### PR TITLE
[SDK 328] Unique time stamp

### PR DESCRIPTION
**Pull request by ZahidZafar**

### CODE REVIEW

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
This PR adds "state" in a static variable, that should not be done. Rework this PR (and probably the TimeHelper class) so that nothing is stored in a static variable.

**Comment by ZahidZafar** [comment]
> This PR adds "state" in a static variable, that should not be done. Rework this PR (and probably the TimeHelper class) so that nothing is stored in a static variable.

To remove static variables and functions that are being used in the whole SDK I have to create a single instance of 'Timehelper' class pass to all classes to use this one.

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
"ToUnixTime" should be reverted back to it's original logic. You are changing it's purpose. It should only create a unix timestamp out of a DateTime. 
Move the new "unique time" logic to "UnixTimeNow".
"UnixTimeNow" should be renamed to "GetUniqueUnixTime".

File: countlyCommon/TestingRelated/RequestTestCases.cs [hunk 1]
```diff
@@ -29,11 +29,7 @@ public void Dispose()
         public void BaseRequestBasic()
         {
             String req = RequestHelper.CreateBaseRequest("a", "b", "c", "d", 123);
-            Assert.Equal("/i?app_key=a&device_id=b&timestamp=123&sdk_version=c&sdk_name=d", req);
-
-            String req2 = RequestHelper.CreateBaseRequest("a", "b", "c", "d");
-            Assert.Contains("/i?app_key=a&device_id=b&timestamp=", req2);
-            Assert.True(req2.EndsWith("&sdk_version=c&sdk_name=d"));
```
[the remark this answers is not in the document]
Comment (ZahidZafar) [reply countlyCommon/TestingRelated/RequestTestCases.cs:36]: This test has been updated in another ticket, but I am updating it back here on your request.

File: countlyCommon/countlyCommon/Helpers/TimeHelper.cs [hunk 2]
```diff
@@ -21,12 +21,59 @@ THE SOFTWARE.
...
+        internal class TimeInstant
+        {
+            public int Dow { get; private set; }
+            public int Hour { get; private set; }
+            public int Timezone { get; private set; }
+            public long Timestamp { get; private set; }
+
+            
+
+            internal TimeInstant()
```
Comment (ArtursKadikis) [inline countlyCommon/countlyCommon/Helpers/TimeHelper.cs:40]: This constructor should not exist

File: countlyCommon/countlyCommon/Helpers/TimeHelper.cs [hunk 3]
```diff
@@ -21,12 +21,59 @@ THE SOFTWARE.
...
+                Dow = (int)dateTime.DayOfWeek;
+                Hour = dateTime.TimeOfDay.Hours;
+
+                TimeSpan ts = dateTime.Subtract(new DateTime(1970, 1, 1));
+                Timestamp = (long)ts.TotalMilliseconds;
+
+                Timezone = (int)TimeZoneInfo.Local.GetUtcOffset(dateTime).TotalMinutes;
+            }
+
+            internal TimeInstant(long timestampInMillis, int hour, int dow)
```
Comment (ArtursKadikis) [inline countlyCommon/countlyCommon/Helpers/TimeHelper.cs:53]: This constructor needs all 4 fields. The timezone is not set bellow and you are already introducing a bug.

File: countlyCommon/countlyCommon/Helpers/TimeHelper.cs [hunk 4]
```diff
@@ -21,12 +21,59 @@ THE SOFTWARE.
...
+            {
+                Dow = dow;
+                Hour = hour;
+                Timestamp = timestampInMillis;
+            }
+
+            internal static TimeInstant Get(long timestampInMillis)
```
Comment (ArtursKadikis) [inline countlyCommon/countlyCommon/Helpers/TimeHelper.cs:60]: This doesn't set a timezone

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
You are again ignoring the required changed. There is no "GetUniqueInstant" call in the "TimeHelper" class. That is the function that should be used.
By ignoring that class you are again introducing another bug by not guaranteing unique timestamps.

Change the timezone data type to "String"

File: countlyCommon/countlyCommon/Helpers/TimeHelper.cs [hunk 5]
```diff
@@ -21,25 +21,88 @@ THE SOFTWARE.
...
+            long calculatedMillis = (long)ts.TotalMilliseconds;
+
+            return calculatedMillis;
         }
 
-        public static long UnixTimeNow()
+        public long GetUniqueUnixTime()
         {
-            return ToUnixTime(DateTime.Now.ToUniversalTime());
+            long calculatedMillis =  ToUnixTime(DateTime.Now.ToUniversalTime());
```
Comment (ArtursKadikis) [inline countlyCommon/countlyCommon/Helpers/TimeHelper.cs:97]: reduntant empty space

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]

**Pull request merged**